Add tests for emptying list, then adding elements

jcshih opened the pull request:
Add tests cases to ensure the linked list's state is properly reset when popping/shifting its last element.

matthewmorgan commented:
@jcshih -- thanks for catching this (especially since that was my `example.js` ;) !!)

I think it could be written so the logic is a little more clear.  In hindsight I think the ternaries were not the best choice. What do you think about this, for example:
```
  shift() {
    if (!this.tail) {
      return undefined;
    }
    const value = this.tail.value;
    if (this.tail.next){
      this.tail = this.tail.next;
      this.tail.next = null
    } else {
      this.tail = null;
      this.head = null;
    }
    return value;
  }
```

Also, we usually like to see one commit per PR if possible.  If you need help squashing the commits please let me know, I'd be happy to assist.

jcshih commented:
@matthewmorgan Thanks for the tips. I agree that conditionals are more readable than ternaries in this case.

I went ahead and refactored and squashed all the changes. I also switched next/prev because it made more sense to me.

matthewmorgan commented:
Thanks @jcshih !
